File: content/manual/v8.2/source/includes/aggregation/aggregation-examples/filtered-subset/full-files/FilteredSubset.kt
package org.example

import com.mongodb.client.model.Aggregates
import com.mongodb.client.model.Filters
import com.mongodb.client.model.Sorts
import com.mongodb.kotlin.client.coroutine.MongoClient
import kotlinx.datetime.LocalDateTime
import kotlinx.serialization.Contextual
import kotlinx.serialization.Serializable
import org.bson.Document
import org.bson.conversions.Bson

// start-data-classes
@Serializable
data class Address(
    val number: Int,
    val street: String,
    val city: String
)

@Serializable
data class Person(
    val personID: String,
    val firstname: String,
    val lastname: String,
    @Contextual val dateOfBirth: LocalDateTime,
    val vocation: String,
    val address: Address,
    val gender: String? = null
)
// end-data-classes

suspend fun main() {
    val uri = "<connection string>"

    MongoClient.create(uri).use { mongoClient ->
        val aggDB = mongoClient.getDatabase("agg_tutorials_db")
        // start-insert-persons
        val persons = aggDB.getCollection<Person>("persons")
        persons.deleteMany(Filters.empty())

        persons.insertMany(
            listOf(
                Person(
                    "6392529400",
                    "Elise",
                    "Smith",
                    LocalDateTime.parse("1972-01-13T09:32:07"),
                    "ENGINEER",
                    Address(5625, "Tipa Circle", "Wojzinmoj")
                ),
                Person(
                    "1723338115",
                    "Olive",
                    "Ranieri",
                    LocalDateTime.parse("1985-05-12T23:14:30"),
                    "ENGINEER",
                    Address(9303, "Mele Circle", "Tobihbo"),
                    "FEMALE"
                ),
                Person(
                    "8732762874",
                    "Toni",
                    "Jones",
                    LocalDateTime.parse("1991-11-23T16:53:56"),
                    "POLITICIAN",
                    Address(1, "High Street", "Upper Abbeywoodington")
                ),
                Person(
                    "7363629563",
                    "Bert",
                    "Gooding",
                    LocalDateTime.parse("1941-04-07T22:11:52"),
                    "FLORIST",
                    Address(13, "Upper Bold Road", "Redringtonville")
                ),
                Person(
                    "1029648329",
                    "Sophie",
                    "Celements",
                    LocalDateTime.parse("1959-07-06T17:35:45"),
                    "ENGINEER",
                    Address(5, "Innings Close", "Basilbridge")
                ),
                Person(
                    "7363626383",
                    "Carl",
                    "Simmons",
                    LocalDateTime.parse("1998-12-26T13:13:55"),
                    "ENGINEER",
                    Address(187, "Hillside Road", "Kenningford")
                )
            )
        )
        // end-insert-persons
        val pipeline = mutableListOf<Bson>()

        // start-match
        pipeline.add(Aggregates.match(Filters.eq(Person::vocation.name, "ENGINEER")))
        // end-match


        // start-sort
        pipeline.add(Aggregates.sort(Sorts.descending(Person::dateOfBirth.name)))
        // end-sort

        // start-limit
        pipeline.add(Aggregates.limit(3))
        // end-limit

        // start-unset
        pipeline.add(Aggregates.unset("_id", Person::address.name))
        // end-unset

        // start-run-agg
        val aggregationResult = persons.aggregate<Document>(pipeline)
        // end-run-agg

        aggregationResult.collect { println(it) }
    }
}

File: content/manual/v8.2/source/includes/aggregation/aggregation-examples/filtered-subset/full-files/filtered_subset.rb
# typed: strict
require "mongo"
require "bson"

uri = "<connection string>"

Mongo::Client.new(uri) do |client|
  agg_db = client.use("agg_tutorials_db")

  # start-insert-persons
  persons = agg_db[:persons]
  persons.delete_many({})

  persons.insert_many(
    [
      {
        person_id: "6392529400",
        firstname: "Elise",
        lastname: "Smith",
        dateofbirth: DateTime.parse("1972-01-13T09:32:07Z"),
        vocation: "ENGINEER",
        address: {
          number: 5625,
          street: "Tipa Circle",
          city: "Wojzinmoj",
        },
      },
      {
        person_id: "1723338115",
        firstname: "Olive",
        lastname: "Ranieri",
        dateofbirth: DateTime.parse("1985-05-12T23:14:30Z"),
        gender: "FEMALE",
        vocation: "ENGINEER",
        address: {
          number: 9303,
          street: "Mele Circle",
          city: "Tobihbo",
        },
      },
      {
        person_id: "8732762874",
        firstname: "Toni",
        lastname: "Jones",
        dateofbirth: DateTime.parse("1991-11-23T16:53:56Z"),
        vocation: "POLITICIAN",
        address: {
          number: 1,
          street: "High Street",
          city: "Upper Abbeywoodington",
        },
      },
      {
        person_id: "7363629563",
        firstname: "Bert",
        lastname: "Gooding",
        dateofbirth: DateTime.parse("1941-04-07T22:11:52Z"),
        vocation: "FLORIST",
        address: {
          number: 13,
          street: "Upper Bold Road",
          city: "Redringtonville",
        },
      },
      {
        person_id: "1029648329",
        firstname: "Sophie",
        lastname: "Celements",
        dateofbirth: DateTime.parse("1959-07-06T17:35:45Z"),
        vocation: "ENGINEER",
        address: {
          number: 5,
          street: "Innings Close",
          city: "Basilbridge",
        },
      },
      {
        person_id: "7363626383",
        firstname: "Carl",
        lastname: "Simmons",
        dateofbirth: DateTime.parse("1998-12-26T13:13:55Z"),
        vocation: "ENGINEER",
        address: {
          number: 187,
          street: "Hillside Road",
          city: "Kenningford",
        },
      },
    ]
  )
  # end-insert-persons

  pipeline = [
    # start-match
    { "$match": { "vocation": "ENGINEER" } },
    # end-match
    # start-sort
    { "$sort": { "dateofbirth": -1 } },
    # end-sort
    # start-limit
    { "$limit": 3 },
    # end-limit
    # start-unset
    { "$unset": ["_id", "address"] },
    # end-unset
  ]

  # start-run-agg
  aggregation_result = persons.aggregate(pipeline)
  # end-run-agg

  aggregation_result.each do |doc|
    puts doc
  end
end
